resolved: remove entry from cache when goodbye packet received

vchillara opened the pull request:
ttl of an incoming mdns goodbye packet is updated to 1 as per RFC6762 10.1. However, in dns_cache_put_positive(), we check for min_ttl <= 0 to remove the record from cache.
As a result, the entries for which goodbye packets are received do not get removed from cache. changed the check to min_ttl <= 1 to account for goodbye packets.

<!-- devel-freezer = {"comment-id":"1628407913","freezing-tag":"v255-rc2"} -->

poettering commented on src/resolve/resolved-dns-cache.c:
```diff
@@ -442,8 +442,9 @@ static int dns_cache_put_positive(
-        if (min_ttl <= 0) {
+        /* New TTL is 0? Delete this specific entry...
+         * In RFC6762 10.1 for Good Bye packet TTL is 1 */
+        if (min_ttl <= 1) {
```
hmm, as I read the RFC we should instead when min_ttl is zero go through all our matching cached entries and set the stored TTL to 1s.

yuwata replied:
How about the following? When `min_ttl == 1`, find matching existing positive cache entries and update its TTL and cache lifetime `DnsCacheItem.until`.

vchillara replied:
It makes sense to update the ttl/until of existing matching cache entries to 1 second when we get a TTL==0.
However, there also needs to be some sort of callback after one second to check if the entry was updated in the meantime, and remove the entry if no update was received.
Without this, the entry lingers in the cache indefinitely until the cache is pruned the next time.
We stumbled into this behavior while implementing #22532, with the aim of notifying clients when a service expires.

vchillara commented on src/resolve/resolved-mdns.c:
```diff
@@ -397,6 +406,17 @@ static int on_mdns_packet(sd_event_source *s, int fd, uint32_t revents, void *us
                                 /* See the section 10.1 of RFC6762 */
                                 rr->ttl = 1;
+
+                                r = sd_event_add_time_relative(
```
dns_cache_lookup() is only called when a packet is sent, which happens indeterminately.
RFC 6762 section 10.1 does mention that the entry should be deleted form the cache 1 second later.
This helps in keeping the cache up-to-date.
Without this, the entry would linger in cache indefinitely until the next packet is sent.
It also helps to notify the client immediately when a service sends a goodbye packet and expires (#22532).

yuwata replied:
"the entry should be deleted form the cache 1 second later" does not mean we really need to remove the entry after 1 sec. But it is enough to drop the entry when it is requested later than 1sec.

In #22532, please call `dns_cache_prune()` before `dns_cache_lookup()` in `mdns_subscriber_lookup_cache()`.

vchillara replied:
Since goodbye packets are unsolicited, and this logic is actually independent of the RFC6762 section 5.2 stuff implemented in #22532, I figured it would make sense to add here.

poettering commented:
hmm, what i don't get: why have a timer to remove these entries from the cache at all? it should suffice to clean this up when we query the cache, and encounter an entry which is now run over. we generally only clean the cache when using it, but not independently by timer. so why should we do so here?

patching the ttl to 1 should mean that if we then query the cache it the record will be used still within that 1s but once the 1s is hit it won't anymore, even if it still exists in the cache.

what am i missing?

vchillara commented:
@poettering thanks for the review. Working on addressing your feedback.
Regarding your question - does https://github.com/systemd/systemd/pull/22751#discussion_r908131303 answer it?

poettering commented:
> @poettering thanks for the review. Working on addressing your feedback. Regarding your question - does [#22751 (comment)](https://github.com/systemd/systemd/pull/22751#discussion_r908131303) answer it?

Not really. Why would it be beneficial to clean up the cache immediately when the timeout runs out? sounds much better to just do so when we deal with the cache anyway. i.e. usually it makes sense to avoid wakeups for nothing, and instead do all work when we are awake anyway...

vchillara commented:
> > @poettering thanks for the review. Working on addressing your feedback. Regarding your question - does [#22751 (comment)](https://github.com/systemd/systemd/pull/22751#discussion_r908131303) answer it?
> 
> Not really. Why would it be beneficial to clean up the cache immediately when the timeout runs out? sounds much better to just do so when we deal with the cache anyway. i.e. usually it makes sense to avoid wakeups for nothing, and instead do all work when we are awake anyway...

I see your point about avoiding unnecessary wakeups, but in the context of browsing services, we would like to keep the client up to date. Since we notify them on addition of services, we also want to notify them when a service expires.
Besides, the logic of removing cache entries with a timer would only be active while there is an active subscriber. 
What do you think?

vchillara commented:
@poettering Added a comment clarifying the need for this.
One thing - in some of the earlier comments on this PR, there was a request to move this functionality to #22532, which we did.
So the question is do we now want it here or in the other PR?

poettering commented on src/resolve/resolved-mdns.c:
```diff
@@ -397,6 +406,20 @@ static int on_mdns_packet(sd_event_source *s, int fd, uint32_t revents, void *us
+                                /* Look at the cache 1 second later and remove stale entries.
+                                 * This is particularly useful to keep service browsers updated on service removal,
+                                 * as there are no other reliable triggers to propogate that info. */
+                                r = sd_event_add_time_relative(
```
so this will create one event source for each incoming mdns packet with ttl set to 0. that is a chance for a DoS, as clients can make us allocate basically unbounded memory, if they send us a ton of these packets within 1s.

Thus the event source object should be stored somewhere (in the scope object?), and if it already exists no new event source should be allocated. morevoer, the handler should check if there are RRs in the cache that should be flushed within less than 1s, and if so should immediately requeue a new event source to cover for that.

That way there's only one event source enqueued at max.

poettering commented on src/resolve/resolved-mdns.c:
```diff
@@ -397,6 +406,20 @@ static int on_mdns_packet(sd_event_source *s, int fd, uint32_t revents, void *us
+                                                USEC_PER_SEC,
+                                                0,
+                                                mdns_goodbye_callback,
+                                                scope);
```
this will result in memory corruption: let's say a scope goes away (because a network interface goes away), but the event source is still queued. it will be called pointing to a scope that doesn't exist anymore. Bad things will then happen.

if you embedd a pointer to the eent source in DnsScope object, then you can also delete the event source whenever the scope goes away.

yuwata commented on src/resolve/resolved-dns-cache.h:
```diff
@@ -58,3 +58,5 @@ bool dns_cache_is_empty(DnsCache *cache);
 
 int dns_cache_export_shared_to_packet(DnsCache *cache, DnsPacket *p, usec_t ts, unsigned max_rr);
+
+bool dns_cache_expiry_in_one_second(DnsCache *c, usec_t t);
```
Missing line break at the end of the file.

yuwata commented on src/resolve/resolved-mdns.c:
```diff
@@ -406,6 +433,22 @@ static int on_mdns_packet(sd_event_source *s, int fd, uint32_t revents, void *us
+                                 * This is particularly useful to keep service browsers updated on service removal,
+                                 * as there are no other reliable triggers to propogate that info. */
+                                if (!scope->mdns_goodbye_event_source) {
+                                        r = sd_event_add_time_relative(
```
Still I cannot follow why we need to actively clear the outdated cache entries...

vchillara replied:
In the context of continuous browsing (#22532), it would be useful to actively clear the outdated cache entries and notify the browser client(s).
To accomplish that, we would eventually add a call to mdns_notify_browsers_unsolicited_updates()
from mdns_goodbye_callback().

yuwata commented:
-> #30096 ?
